src/events/NeverminedEvent.ts

```diff
@@ -0,0 +1,74 @@
+import ContractBase from '../keeper/contracts/ContractBase'
+
+export interface EventOptions {
+    methodName?: string
+    eventName?: string
+    filterJsonRpc?: {}
+    filterSubgraph?: {}
+    result?: {}
+    fromBlock?: number | string
+    toBlock?: number | string
+}
+
+export interface EventEmitter {
+    subscribe: Function
+    unsubscribe: Function
+}
+
+export interface ContractEventSubscription {
+    unsubscribe: () => void
+}
+
+export type EventResult = Promise<Array<any>>
+
+export abstract class NeverminedEvent {
+    protected eventEmitter: EventEmitter
+    protected contract: ContractBase = null
+    public abstract getEventData(options: EventOptions): EventResult
+    public abstract getPastEvents(options: EventOptions): EventResult
+    public abstract getBlockNumber(...args: any[]): Promise<number>
+
+    protected constructor(contract: ContractBase, eventEmitter: EventEmitter) {
+        this.contract = contract
+        this.eventEmitter = eventEmitter
+    }
+
+    public subscribe(
+        callback: (events: EventResult[]) => void,
+        options: EventOptions
+    ): ContractEventSubscription {
+        const onEvent = async () => {
+            const events = await this.getEventData(options)
+            if (events.length) {
+                callback(events)
+            }
+        }
+        this.eventEmitter.subscribe(onEvent, () => this.getBlockNumber())
+        return {
+            unsubscribe: () => this.eventEmitter.unsubscribe(onEvent)
+        }
+    }
+
+    public async once(
+        callback?: (events: EventResult[]) => void,
+        options?: EventOptions
+    ) {
+        // Check if the event already happened and return that instead
+        // before subscribing
+        const events = await this.getPastEvents(options)
```

Comment:
👍  will be of great use

Reply:
yes. Although you should always prefer `once` to `getPastEvents` because it works differently with the graph because the indexer node always lags behind a bit